File: frontend/src/api/sourceImport.ts
// Shared client for the "source import" assistants — a live third-party
// system (Monica, issue #549) or an uploaded database (Meerkat, issue #550)
// pulled into a background session, reviewed with a loss report, then
// confirmed through the same staged preview→confirm contract as the file
// imports. Everything here is source-agnostic and parameterised by a
// `basePath`; the Monica-specific connect/fetch calls live in monicaImport.ts.
//
// Phase and category vocabularies mirror backend/models/monica_import.go and
// backend/services/import_source.go by hand — there is no dynamic type-list
// endpoint (CLAUDE.md frontend trap #4). Keep them in sync.
import { API_BASE_URL, apiFetch, getAuthHeaders, parseErrorResponse } from './client';
import type { ImportRowPreview } from './import';

export type SourceImportPhase =
  | 'connecting'
  | 'parsing_database'
  | 'mapping'
  | 'fetching_contacts'
  | 'fetching_activities'
  | 'fetching_notes'
  | 'fetching_reminders'
  | 'fetching_extras'
  | 'fetching_relationships'
  | 'building_preview'
  | 'ready'
  | 'importing'
  | 'importing_photos'
  | 'done'
  | 'failed'
  | 'cancelled';

// The review step's per-contact decision. Mirrors backend RowImportAction's
// `oneof=skip add update`.
export type RowSourceAction = 'add' | 'skip' | 'update';

export type SourceImportIssueCategory =
  | 'unsupported'
  | 'lossy'
  | 'transformed'
  | 'invalid'
  | 'skipped';

// One entry of the pre-confirm loss report (the issue #442 record/field/
// category shape from services.ImportIssue).
export interface SourceImportIssue {
  record: string;
  field: string;
  category: SourceImportIssueCategory;
  message: string;
}

// Per-contact tally of graph entities the import will create for that row.
export interface SourceRelatedCounts {
  activities: number;
  notes: number;
  reminders: number;
  relationships: number;
  gifts: number;
}

export interface SourceImportRowPreview extends ImportRowPreview {
  related: SourceRelatedCounts;
  has_photo: boolean;
}

export interface SourceImportResult {
  total_processed: number;
  created: number;
  updated: number;
  skipped: number;
  errors: string[];
  relationships_created: number;
  notes_created: number;
  activities_created: number;
  reminders_created: number;
  gifts_created: number;
  custom_fields_created: number;
  photos_queued: number;
  photos_saved: number;
  photos_failed: number;
}

export interface SourceImportStatus {
  session_id: string;
  phase: SourceImportPhase;
  phase_done: number;
  phase_total: number;
  error?: string;
  result?: SourceImportResult;
}

export interface SourceImportPreviewResponse {
  session_id: string;
  rows: SourceImportRowPreview[];
  total_rows: number;
  valid_rows: number;
  duplicate_count: number;
  error_count: number;
  totals: SourceRelatedCounts;
  loss_report: SourceImportIssue[];
}

export interface RowSourceActionInput {
  row_index: number;
  action: RowSourceAction;
}

async function getJSON<T>(path: string): Promise<T> {
  const response = await apiFetch(`${API_BASE_URL}${path}`, { headers: getAuthHeaders() });
  if (!response.ok) throw await parseErrorResponse(response);
  return response.json();
}

export function getSourceImportStatus(
  basePath: string,
  sessionId: string,
): Promise<SourceImportStatus> {
  return getJSON(`${basePath}/status?session_id=${encodeURIComponent(sessionId)}`);
}

export function getSourceImportPreview(
  basePath: string,
  sessionId: string,
): Promise<SourceImportPreviewResponse> {
  return getJSON(`${basePath}/preview?session_id=${encodeURIComponent(sessionId)}`);
}

// confirmSourceImport starts the import in the background — the endpoint
// replies 202 with no result body. The caller polls getSourceImportStatus
// until phase "done" and reads the summary from status.result.
export async function confirmSourceImport(
  basePath: string,
  sessionId: string,
  actions: RowSourceActionInput[],
): Promise<void> {
  const response = await apiFetch(`${API_BASE_URL}${basePath}/confirm`, {
    method: 'POST',
    headers: getAuthHeaders(),
    body: JSON.stringify({ session_id: sessionId, actions }),
  });
  if (!response.ok) throw await parseErrorResponse(response);
}

export async function cancelSourceImport(basePath: string, sessionId: string): Promise<void> {
  // Best-effort: a failed cancel just leaves the session to expire on its own.
  await apiFetch(`${API_BASE_URL}${basePath}/cancel?session_id=${encodeURIComponent(sessionId)}`, {
    method: 'POST',
    headers: getAuthHeaders(),
  }).catch(() => undefined);
}

File: backend/models/source_import.go
package models

// Shared DTOs for the "source import" assistants — a live third-party system
// (Monica, issue #549) or an uploaded database (Meerkat, issue #550) pulled
// into a background session, reviewed with a loss report, then confirmed
// through the shared engine. Only the acquisition step differs between
// sources; everything from the review step on is this file.
//
// The phase and category vocabularies are mirrored by hand in
// frontend/src/api/sourceImport.ts and backend/openapi.yaml — there is no
// dynamic type-list endpoint by design (CLAUDE.md frontend trap #4).

// Source-import wizard phases, in rough order. Terminal: done / failed /
// cancelled. The fetch-* phases are Monica-specific; parsing_database /
// mapping are Meerkat-specific; the rest are shared.
const (
	SourceImportPhaseConnecting            = "connecting"
	SourceImportPhaseParsingDatabase       = "parsing_database"
	SourceImportPhaseMapping               = "mapping"
	SourceImportPhaseFetchingContacts      = "fetching_contacts"
	SourceImportPhaseFetchingActivities    = "fetching_activities"
	SourceImportPhaseFetchingNotes         = "fetching_notes"
	SourceImportPhaseFetchingReminders     = "fetching_reminders"
	SourceImportPhaseFetchingExtras        = "fetching_extras"
	SourceImportPhaseFetchingRelationships = "fetching_relationships"
	SourceImportPhaseBuildingPreview       = "building_preview"
	SourceImportPhaseReady                 = "ready"
	SourceImportPhaseImporting             = "importing"
	SourceImportPhaseImportingPhotos       = "importing_photos"
	SourceImportPhaseDone                  = "done"
	SourceImportPhaseFailed                = "failed"
	SourceImportPhaseCancelled             = "cancelled" // an in-flight import was cancelled; the transaction rolled back
)

// SourceImportStatus is the poll payload for the fetch/import progress bar.
type SourceImportStatus struct {
	SessionID  string              `json:"session_id"`
	Phase      string              `json:"phase"`
	PhaseDone  int                 `json:"phase_done"`
	PhaseTotal int                 `json:"phase_total"`
	Error      string              `json:"error,omitempty"`
	Result     *SourceImportResult `json:"result,omitempty"` // set from phase "importing"/"importing_photos" onwards
}

// SourceRelatedCounts is the per-contact tally of graph entities the import
// will create for that contact, shown as chips in the review step so it never
// promises more than the confirm will produce. Notes already folds in the
// source's "extra" record kinds (Monica tasks/debts; nothing extra for
// Meerkat); Activities already folds in Monica's logged calls.
type SourceRelatedCounts struct {
	Activities    int `json:"activities"`
	Notes         int `json:"notes"`
	Reminders     int `json:"reminders"`
	Relationships int `json:"relationships"`
	Gifts         int `json:"gifts"`
}

// SourceImportRowPreview is one contact row of the review step: the shared
// ImportRowPreview (parsed fields, validation, duplicate match, merge diff)
// plus source-import context.
type SourceImportRowPreview struct {
	ImportRowPreview
	Related  SourceRelatedCounts `json:"related"`
	HasPhoto bool                `json:"has_photo"`
}

// SourceImportIssue is one entry of the loss report shown before confirm —
// the #442 record/field/category shape from services.ImportIssue.
type SourceImportIssue struct {
	Record   string `json:"record"`
	Field    string `json:"field"`
	Category string `json:"category"`
	Message  string `json:"message"`
}

// SourceImportPreviewResponse is the full review-step payload.
type SourceImportPreviewResponse struct {
	SessionID      string                   `json:"session_id"`
	Rows           []SourceImportRowPreview `json:"rows"`
	TotalRows      int                      `json:"total_rows"`
	ValidRows      int                      `json:"valid_rows"`
	DuplicateCount int                      `json:"duplicate_count"`
	ErrorCount     int                      `json:"error_count"`
	Totals         SourceRelatedCounts      `json:"totals"`
	// LossReport is every mapping issue, so the user sees what the import
	// cannot carry before they commit (issue #442). Never omitempty: an empty
	// import still returns [] so the client can render "nothing lost".
	LossReport []SourceImportIssue `json:"loss_report"`
}

// SourceImportConfirmRequest executes the import with the per-contact
// decisions from the review step. Reuses RowImportAction (skip|add|update);
// "update" merges the mapped contact into the row's DuplicateMatch.
type SourceImportConfirmRequest struct {
	SessionID string            `json:"session_id" validate:"required"`
	Actions   []RowImportAction `json:"actions" validate:"required,dive"`
}

// SourceImportResult is the outcome summary. Embeds the shared ImportResult
// (total/created/updated/skipped/errors) and adds the graph-entity counts and
// (Monica only) the deferred photo progress — Meerkat leaves the photo
// counters at zero.
type SourceImportResult struct {
	ImportResult
	RelationshipsCreated int `json:"relationships_created"`
	NotesCreated         int `json:"notes_created"`
	ActivitiesCreated    int `json:"activities_created"`
	RemindersCreated     int `json:"reminders_created"`
	GiftsCreated         int `json:"gifts_created"`
	CustomFieldsCreated  int `json:"custom_fields_created"`
	PhotosQueued         int `json:"photos_queued"`
	PhotosSaved          int `json:"photos_saved"`
	PhotosFailed         int `json:"photos_failed"`
}
